lib/spack/spack/spec.py

```diff
@@ -1740,17 +1770,30 @@ def to_dict(self, hash=ht.dag_hash):
 
         """
         node_list = []
+        hash_list = []
         for s in self.traverse(order='pre', deptype=hash.deptype):
-            node_list.append(s.node_dict_with_hashes(hash))
-
-        return syaml.syaml_dict([('spec', node_list)])
+            spec_hash = s.node_dict_with_hashes(hash)[hash.attr]
+            if spec_hash not in hash_list:
+                node_list.append(s.node_dict_with_hashes(hash))
+                hash_list.append(spec_hash)
+            if s.build_spec is not s:
+                build_spec_list = s.build_spec.to_dict()['spec']['nodes']
+                for node in build_spec_list:
+                    node_hash = node[hash.attr]
+                    if node_hash not in hash_list:
+                        node_list.append(node)
+                        hash_list.append(node_hash)
```

Comment:
Would this be simpler using a set for all the dependencies instead of having to track whether they're already present?

Reply:
Yeah makes sense.

Reply:
Changed to `hash_set`.